Refactor Zone Creation `[AARD-2071]`

AlexD717 opened the pull request:
## Task
Refactor the way zones are created in synthesis.
<!--
Please include any relevant Jira ticket ID(s) at the end of the PR title, in the form AARD-xxxx, where "AARD" is Jira project.
Include the same Jira ticket ID(s) in this section.
-->

AARD-2071

<!--
Provide a brief description of what the task was here.
-->

## Symptom
There is currently a lot of duplicate code in the way Scoring Zones and Protected Zones are set up. If we decide that we want to add more zones later, ex for a loading station zone (AARD-2072) there would be even more duplicate code.
<!--
How does the problem manifest itself?
Describe the problem as seen by the user, or by the caller or the code if it's not directly visible to the user.

Note: "Symptom" can include new product use cases, not just bugs.
-->

## Solution
Move the code that is the same for all zones, such as the name, and mainly gizmo configuration into an outside component, making it easier to add more zones later in the future.
<!--
How did you fix the problem/symptom?
Explain your approach and reasoning for choosing this solution.
-->

## Verification
- All current zone related functionality still works
<!--
How did you test and verify your changes were correct?
List steps taken, tests/added/updated, and any manual verification done that should be replicated in review.
-->

---

Before merging, ensure the following criteria are met:

- [x] All acceptance criteria outlined in the ticket are met.
- [x] Necessary test cases have been added and updated.
- [x] A feature toggle or safe disable path has been added (if applicable).
- [x] User-facing polish:
  - Ask: *"Is this ready-looking?"*
- [x] Cross-linking between Jira and GitHub:
  - PR links to the relevant Jira issue.
  - Jira ticket has a comment referencing this PR.

AlexD717 committed:
Simple Refactor

AlexD717 committed:
Merge branch 'dev' into alexey/2071/refactor-zone-creation

AlexD717 committed:
Comments

AlexD717 committed:
Formatting

rutmanz commented on fission/src/ui/panels/configuring/assembly-config/interfaces/scoring/ProtectedZoneConfigInterface.tsx:
```diff
@@ -308,26 +90,22 @@ const ZoneConfigInterface: React.FC<ZoneConfigProps> = ({ selectedField, selecte
                     value={activeDuring}
                     input={<OutlinedInput label="Contact Type" />}
-                    renderValue={selected => selected.join(", ")}
+                    renderValue={selected => (selected as MatchModeType[]).join(", ")}
```
why is there a cast here

rutmanz commented on fission/src/ui/panels/configuring/assembly-config/interfaces/scoring/ManageScoringZonesInterface.tsx:
```diff
@@ -1,140 +1,42 @@
     selectZone: (zone: ScoringZonePreferences) => void
 }
 
 const ManageZonesInterface: React.FC<ScoringZonesProps> = ({ selectedField, initialZones, selectZone }) => {
```
`ManageScoringZonesInterface`?

rutmanz commented on fission/src/ui/panels/configuring/assembly-config/interfaces/scoring/ManageProtectedZonesInterface.tsx:
```diff
@@ -1,137 +1,44 @@
     selectZone: (zone: ProtectedZonePreferences) => void
 }
 
 const ManageZonesInterface: React.FC<ProtectedZonesProps> = ({ selectedField, initialZones, selectZone }) => {
```
`ManageProtectedZonesInterface`?

rutmanz commented on fission/src/ui/panels/configuring/assembly-config/interfaces/scoring/ScoringZoneConfigInterface.tsx:
```diff
@@ -105,197 +23,33 @@ interface ZoneConfigProps {
 }
 
 const ZoneConfigInterface: React.FC<ZoneConfigProps> = ({ selectedField, selectedZone, saveAllZones }) => {
```
Scoring

rutmanz commented on fission/src/ui/panels/configuring/assembly-config/interfaces/scoring/ScoringZoneConfigInterface.tsx:
```diff
@@ -105,197 +23,33 @@ interface ZoneConfigProps {
+            selectedZone={selectedZone}
+            attachAndPersistZone={attachAndPersistZone}
+            applyExtrasOnSave={zone => {
+                zone.points = zone.points ?? selectedZone.points
```
why are we only updating the points if they were previously null?

rutmanz reviewed (changes requested):
- Turning on persistent points for a scoring zone or either dropdown for protected zones seems to reset the position and scale
- The red material for scoring zones get blocked by objects in front but the blue does not (I think it's better if they're both occluded because it makes it easier to position things)
- It's not strictly in scope but the buttons being down below in the main menu is kind of weird and it might make sense to fix as part of this

AlexD717 committed:
PR Feedback

AlexD717 committed:
Proper Material & Unnecessary Button Removed

AlexD717 committed:
Formatting

rutmanz reviewed (commented):
otherwise lgtm

rutmanz commented on fission/src/mirabuf/MirabufSceneObject.ts:
```diff
@@ -600,22 +600,28 @@ class MirabufSceneObject extends SceneObject implements ContextSupplier {
+			if (parts) {
+				const editor = new FieldMiraEditor(parts)
+				editor.setUserData("devtool:scoring_zones", this._fieldPreferences.scoringZones)
+			}
```
They already persist in the preferences system. I don't think that users should implicitly be updating their local cache files like this

AlexD717 committed:
Remove Unnecessary Cache

rutmanz commented on fission/src/mirabuf/MirabufSceneObject.ts:
```diff
@@ -814,6 +814,15 @@ class MirabufSceneObject extends SceneObject implements ContextSupplier {
+                        editor.setUserData("devtool:scoring_zones", this._fieldPreferences.scoringZones)
+                    }
+                    if (this._fieldPreferences.spawnLocations?.hasConfiguredLocations) {
+                        editor.setUserData("devtool:spawn_locations", this._fieldPreferences.spawnLocations)
```
is this part also not needed for the same reason?

AlexD717 committed:
Removed More Unnecessary Code

AlexD717 committed:
Revert "Removed More Unnecessary Code"

This reverts commit e9dbaece7a7e9f5de745bec1b98cf35c71462498.

AlexD717 committed:
Merge branch 'dev' into alexey/2071/refactor-zone-creation

AlexD717 committed:
Reapply "Removed More Unnecessary Code"

This reverts commit a94204a05335a3fbe496ebc9215f4759d5784a76.

BrandonPacewic committed:
fix: correct zone refactor regressions

  - Revert default scoring zone points from 5 to 0
  - Fix depthTest set to true instead of false, breaking gizmo visibility
  - Fix new zones never being added to the zones array before selectZone()
  - Restore back arrow button in ConfigureScoringZonesInterface (was removed)
  - Add back arrow button to ConfigureProtectedZonesInterface (was never present)
  - Fix parseInt fallback from 1 to 0 so zones can be set to 0 points
  - Use stable composite key instead of array index in zone list
  - Add emptyLabel prop to ManageZonesBase with type-specific defaults
  - Constrain ManageZonesBase TZone generic to BaseZonePreferences

BrandonPacewic committed:
Merge `dev`

* dev: (206 commits)
  fix: biome lint warnings
  feat: import unstable mira files with warnings
  revert: `2fd4c67aae0be43b473ed0e5b5f2d90f17ecfcf6`
  feat: handle zero triangles in `createConcaveShapeSettingsFromPart`
  doc: add current status to in development features
  docs: update manual exporter install tutorial
  docs: tutorial content updates
  fix: spelling, grammer, and formatting corrections
  Update tutorials/CommandPalette.md
  Update tutorials/SpawnAsset.md
  Update tutorials/SpawnAsset.md
  Update tutorials/Multiplayer.md
  Update tutorials/MatchMode.md
  Update tutorials/MatchMode.md
  Update tutorials/MatchMode.md
  fix(exporter): fixed weird python std throw error
  fix: final readme fixes
  fix: remove commited config file
  fix: migrate biome config
  fix: remove prints
  ...

BrandonPacewic committed:
fix: build isuses

BrandonPacewic committed:
fix: biome fmt issues

BrandonPacewic committed:
fix: revert `fission/src/util/TypeConversions.ts` formatting

BrandonPacewic committed:
fix: don't add new zone to array before navigating to config

    Adding the new zone to `fieldPrefs.scoringZones` and calling
    `updateScoringZones()` before `selectZone()` caused a
    `ScoringZoneSceneObject`/`ProtectedZoneSceneObject` to be rendered at
    the zone's delta position before `postGizmoCreation` could remove it,
    resulting in the zone appearing at the wrong position on creation.

    The zone is added to preferences via `attachAndPersistZone` on save,
    and `ConfigurationSavedEvent` fires on back-button navigation, so the
    zone is always persisted correctly without needing to be in the array
    beforehand.

BrandonPacewic reviewed (approved):
Added a few polish changes and fixed a couple outstanding issues.

Thanks for doing this. I always love a good code refactor!
